tests/tasty/Arbitrary.hs

```diff
@@ -155,3 +174,109 @@ instance Arbitrary Constant where
+
+-- | Base environment with no declared properties; do not add an Arbitrary
+-- instance to this and instead use newtypes.
+newtype Env = Env { unEnv :: M.HashMap Symbol SortedReft }
+  deriving (Eq, Show, Generic)
+
+-- | Env without anf vars.
+newtype NoAnfEnv = NoAnfEnv { unNoAnfEnv :: Env }
+  deriving (Eq, Show, Generic)
+instance Arbitrary NoAnfEnv where
+  arbitrary = NoAnfEnv <$> (arbitraryEnv 4 gen)
+    where
+      -- | Note that this relies on the property that the Arbitrary instance for
+      -- Symbol cannot create lq_anf$ vars.
+      symGen = const arbitrary
+      reftGen = const arbitrary
+      gen = fmap pure . (pairGen symGen reftGen)
+
+pairGen :: (Int -> Gen a) -> (Int -> Gen b) -> Int -> Gen (a, b)
+pairGen aGen bGen = \i -> do
+  a <- aGen i
+  b <- bGen i
+  pure (a, b)
+
+-- | Env with anf vars that do not reference further anf vars.
+newtype FlatAnfEnv = FlatAnfEnv { unFlatAnfEnv :: Env }
+  deriving (Eq, Show, Generic)
+instance Arbitrary FlatAnfEnv where
+  arbitrary = FlatAnfEnv <$> (arbitraryEnv 4 gen)
+    where
+      symGen = const arbitrary
+      anfGen' = anfGen symGen (const arbitrary)
+      gen = finalAnfGen (fmap pure . anfGen')
+
+anfGen :: (Int -> Gen AnfSymbol) -> (Int -> Gen SortedReft) -> Int -> Gen (Symbol, SortedReft)
+anfGen symGen sreftGen = \i -> do
+  sym <- unAnfSymbol <$> symGen i
+  sreft <- sreftGen i
+  pure $ (sym, sreft)
+
+finalAnfGen :: (Int -> Gen [(Symbol, SortedReft)]) -> Int -> Gen [(Symbol, SortedReft)]
+finalAnfGen anfsGen = \i -> do
+  anfs <- anfsGen i
+  case anfs of
+    [] -> pure []
+    ((penultimateSym, _):_) -> do
+      ultimateAnf <- do
+        sym <- arbitrary
+        (,) <$> arbitrary {- symbol -} <*> (
+          RR <$> arbitrary {- sort -} <*> (
+            pure $ reft sym (PAtom Eq (EVar sym) (EVar penultimateSym))))
+      pure $ ultimateAnf : anfs
+
+-- | Create an arbitrary env up to size k with the given generator for Symbols
+-- and SortedRefts
+arbitraryEnv :: Int -> (Int -> Gen [(Symbol, SortedReft)]) -> Gen Env
+arbitraryEnv k gen = Env . M.fromList <$> (choose (0, k) >>= gen)
+
+-- | Env with anf vars that form a list of references.
+newtype ChainedAnfEnv = ChainedAnfEnv { unChainedAnfEnv :: Env }
+  deriving (Eq, Show, Generic)
+instance Arbitrary ChainedAnfEnv where
+  arbitrary = ChainedAnfEnv <$> (arbitraryEnv 8 gen)
+    where
+      gen = finalAnfGen $ chainedAnfGen anfSymNGen
+
+chainedAnfGen :: (Int -> Gen AnfSymbol) -> Int -> Gen [(Symbol, SortedReft)]
+chainedAnfGen symGen = \i -> do
+  syms <- fmap unAnfSymbol <$> (for [0..i] symGen)
+  let symPairs :: [(Symbol, Symbol)]
+      symPairs = pairs syms
+      anfPairs :: Gen [(Symbol, SortedReft)]
+      anfPairs = traverse (\(sym, prevSym) -> do
+                              otherSym <- arbitrary
+                              (sym,) <$> (RR <$> arbitrary <*> (reft otherSym <$> (PAtom Eq (EVar otherSym) <$> (arbitraryExprInvolving prevSym i))))) symPairs
```

Comment:
This line is a bit long. I think there's a flipped variant of `traverse` called `for` if that helps tidying the code.

Reply:
Changed.